### PR TITLE
Update fabric link monitoring plan.

### PR DESCRIPTION
**Pull request by jfeng-arista**
Update "Fabric port support on Sonic" HLD

PRs related to updated sections are:
<table>
  <tr>
    <th>Repo</th>
    <th>PR title</th>
    <th>State</th>
    <th>Dependency</th>
  </tr>

  <tr>
    <td>SONiC</td>
    <td><a href="https://github.com/sonic-net/sonic-mgmt/pull/5678">#5678 Update fabric test plan link</a></td>
    <td>Open</td>
    <td>No</td>
  </tr>

  <tr>
    <td>sonic-utilities</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2528">#2528 Add "show fabric reachability" command.</a></td>
    <td>Merged</td>
    <td>No</db>
  </tr>

  <tr>
    <td>sonic-swss</td>
     <td><a href="https://github.com/sonic-net/sonic-swss/pull/2652">#2611 Revert "Revert "[voq][chassis]Add show fabric counters port/queue com</a></td>
    <td>Merged</td>
    <td><a href="https://github.com/sonic-net/sonic-swss/pull/2611">2611</a></td>
  </tf>

 <tr>
    <td>sonic-utilities</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2499">#2499 Add asic id for linecards so "show fabric counters queue/port" can work</a></td>
    <td>Merged</td>
    <td>No</td>
  <tr>
    <td>sonic-swss</td>
     <td><a href="https://github.com/sonic-net/sonic-swss/pull/2522">#2522 Add show fabric counters port/queue commands.</a></td>
    <td>Merged</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2499">2499</a></td>
  </tf>

 <tr>
    <td>sonic-utilities</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2717">#2717 Add fabric_ns to the ns_list when display_option is DISPLAY_ALL</a></td>
    <td>Merged</td>
    <td>No</td>
  </tr>

  <tr>
    <td>sonic-mgmt</td>
    <td><a href="https://github.com/sonic-net/sonic-mgmt/pull/6620">#6620 Add tests to check fabric link status and reachability</a></td>
    <td>Merged</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2717">2717</a></td>
  </tr>

  <tr>
    <td>sonic-buildimage</td>
    <td><a href="https://github.com/sonic-net/sonic-buildimage/pull/14170 ">#14170 Add support data for fabric monitoring in CONFIG_DB</a></td>
    <td>Merged</td>
    <td>No</td>
  </tr>

  <tr>
    <td>sonic-utilities</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2730 ">#2730 Add "config fabric port ..." commands and tests</a></td>
    <td>Merged</td>
    <td><a href="https://github.com/sonic-net/sonic-buildimage/pull/14170 ">14170</a></td>
  </tr>

  <tr>
    <td>sonic-buildimage</td>
    <td><a href="https://github.com/sonic-net/sonic-buildimage/pull/14282 ">#14282 Add Yang model for the fabric configs</a></td>
    <td>Merged</td>
    <td><a href="https://github.com/sonic-net/sonic-buildimage/pull/14170 ">14170</a></td>
  </tr>

<tr>
    <td>sonic-buildimage</td>
    <td><a href="https://github.com/sonic-net/sonic-buildimage/pull/14390 ">#14390 Add Yang model for FABRIC_MONITOR|FABRIC_MONITOR_DATA</a></td>
    <td>Merged</td>
    <td><a href="https://github.com/sonic-net/sonic-buildimage/pull/14170 ">14170</a></td>
  </tr>

<tr>
    <td>sonic-swss</td>
    <td><a href="https://github.com/sonic-net/sonic-swss/pull/2862 ">#2862 Remove fabric queue counters.</a></td>
    <td>Merged</td>
    <td>No</td>
  </tr>

<tr>
    <td>sonic-swss-common</td>
    <td><a href="https://github.com/sonic-net/sonic-swss-common/pull/808 ">#808 Add fabric monitoring tables definitions. </a></td>
    <td>Merged</td>
    <td><a href="https://github.com/sonic-net/sonic-buildimage/pull/14170 ">14170</a></td>
  </tr>

<tr>
    <td>sonic-swss</td>
    <td><a href="https://github.com/sonic-net/sonic-swss/pull/2920">#2920 Add support for fabric monitor daemon. </a></td>
    <td>Open</td>
    <td><a href="https://github.com/sonic-net/sonic-swss-common/pull/808 ">808</a></td>
  </tr>

<tr>
    <td>sonic-swss</td>
    <td>#TBD Start fabric manager daemon in swss containers </a></td>
    <td>TBD</td>
    <td><a href="https://github.com/sonic-net/sonic-swss-common/pull/808 ">808</a></td>
  </tr>

<tr>
    <td>sonic-mgmt</td>
    <td>#TBD Add tests for fabric manager cli and database </a></td>
    <td>TBD</td>
    <td><a href="https://github.com/sonic-net/sonic-swss-common/pull/808 ">808</a></td>
  </tr>
<tr>

<tr>
    <td>sonic-utilities</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2932">#2932 Add a flag of the fabric monitor feature in CONFIG_DB </a></td>
    <td>open</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2730 ">2730</a></td>
  </tr>
<tr>

<tr>
    <td>sonic-swss</td>
    <td>#TBD Add fabric manager daemon handling for fabric monitor feature enable/disable flag</a></td>
    <td>TBD</td>
    <td><a href="https://github.com/sonic-net/sonic-utilities/pull/2730 ">2730</a></td>
  </tr>

<tr>
    <td>More PRs coming </td>
    <td>TBD</td>
    <td>TBD</td>
    <td>TBD</td>
  </tr>
</table>

### CODE REVIEW

**Comment by kenneth-arista** [comment]
@arlakshm for awareness

File: doc/voq/fabric.md [hunk 1]
```diff
@@ -171,6 +172,85 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
 
 In a later phase, a `show fabric status` command will be added to show the remote switch ID and link ID for each fabric link of an ASIC. This will be obtained from the SAI_PORT_ATTR_FABRIC_REACHABILITY port attribute of the fabric port. Note that for fabric links that do not have a link partner because of the configuration of the chassis, this will show the status as `down`. The status will also be `down` for fabric links that are down due to some other physical error. To identify links that are down due to error vs links that are not expected to be up because of the chassis connectivity, we need to build up a list of expected fabric connectivity for each ASIC. This can be computed ahead of time based on the vendor configuration and populated in the minigraph. This will be implemented in a later phase.
```
Comment (rlhui) [inline doc/voq/fabric.md:173]: Please add if this is applicable to all line cards and sup, or not.

File: doc/voq/fabric.md [hunk 2]
```diff
@@ -171,6 +172,85 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+The design of fabric link monitor is intentionally scoped to use local component state such as information local to a linecard or information local to a supervisor. This design simplifies the need for inter-component communication.
+
+### 2.8.1 Monitor Fabric Link Status
+
+Unhealthy fabric links may lead to traffic drops. Fabric link monitoring is an important tool to minimize traffic loss. The fabric link monitor algorithm monitors fabric link status and takes links out of service if one or more criteria are true. This is needed on both fabric ASICs and forwarding ASICs.
+
+#### 2.8.1.1 Cli commands
+
+```
+> config fabric port set [port_id] isolate
```
Comment (arlakshm) [inline doc/voq/fabric.md:188]: can you change the command `config fabric port isolate <port_id>`

File: doc/voq/fabric.md [hunk 3]
```diff
@@ -171,6 +172,85 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+```
+
+```
+> config fabric port remove [port_id] isolate
```
Comment (arlakshm) [inline doc/voq/fabric.md:192]: can you change the command `config fabric port unisolate <port_id>`

File: doc/voq/fabric.md [hunk 4]
```diff
@@ -171,6 +172,85 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+
+#### 2.8.1.2 Fabric link monitoring criteria
+
+The fabric link monitoring algorithm checks two type of errors on a link: crc errors and uncorrectable errors. 
+
+The criteria can be extended to include checking other errors later.
+
+#### 2.8.1.2 Monitoring algorithm
+
+Instead of reacting to the counter changes, Orchagent adds a new poller and periodically polls status of all fabric links. By default, the total number of received cells, cells with crc errors, cells with uncorrectable errors are fetched from all serdes links periodically and the error rates are caculated using these numbers. If any one of the error rate is above the threshold for a number of consecutive polls, the link is identified as a unhealthy link. Then the link is automatically isolated to not distribute traffic.
```
Comment (arlakshm) [inline doc/voq/fabric.md:205]: In the above section CLI command is defined to isolate a fabric link, in this section it is mentioned `The link is automatically isolated to not distribute traffic. ` If the links isolated automatically, can elaborate  the use case of the isolate CLIs ?

File: doc/voq/fabric.md [hunk 5]
```diff
@@ -171,6 +172,85 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+
+### 2.8.2 Monitor Fabric Capacity
+
+When the fabric link monitoring feature is enabled, fabric links may not be operational in a system due to link down, or link isolation by the monitoring algorithm. As a result, the effective capacity of total fabric links may be less than required bandwith, and lead to performance degradation. Implementing a capacity monitoring algorithm in Orchagent will be useful to alert capacity changes.
+
+
+#### 2.8.2.1 Cli command
+
+```
+> config fabric monitor capacity threshold <50-100>
```
Comment (arlakshm) [inline doc/voq/fabric.md:215]: Please add if the command is applicable to linecard or supervisor or both

File: doc/voq/fabric.md (hunk 1 above)
Comment (jfeng-arista) [reply doc/voq/fabric.md:173]: The commands are for both linecards and fabric cards on sup. I updated the document with this information as well.

File: doc/voq/fabric.md [hunk 6]
```diff
@@ -171,6 +172,85 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+### 2.8.3 Monitor Traffic on Fabric Links
+
+Monitoring traffic on fabric links is another important tool to diagnose fabric hardware issues. It is useful to identify when traffic is unbalanced among fabric links which are connected to the same forwarding ASIC. It can also help identify miswired links.
+
+#### 2.8.3.1 Cli command
+
+The following proposed CLI is used to show the traffic among fabric links on both fabric ASICs and forwarding ASICs.
+
+```
+> show fabric counters rate
```
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply doc/voq/fabric.md:246]: I updated document with this . thank you

File: doc/voq/fabric.md (hunk 2 above)
Comment (jfeng-arista) [reply doc/voq/fabric.md:188]: I updated document  . thank you

File: doc/voq/fabric.md (hunk 4 above)
Comment (jfeng-arista) [reply doc/voq/fabric.md:205]: Great question.  The fabric inks are isolated automatically by the algorithm if they are unhealthy.  

The two commands are kind of additional debug tool.  The commands can be used to manually isolate and unisolate a fabric link and can help us debug on the system as well as force isolate a fabric link.

File: doc/voq/fabric.md (hunk 5 above)
Comment (jfeng-arista) [reply doc/voq/fabric.md:215]: I updated document . thank you

File: doc/voq/fabric.md [hunk 7]
```diff
@@ -171,6 +172,85 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+         Links
+-----    ------      --------     ----     ----------   ---------
+0        110         112          98       None          Never
+1        112         112          100      None          Never  
+....
+```
+
+#### 2.8.2.2 Monitoring algorithm
+
+Orchagent will track the total number of fabric links that are isolated. Once the number of total operational fabric links is below a configured threshold, alert users with a system log.
```
[the remark this answers is not in the document]
Comment (jfeng-arista) [reply doc/voq/fabric.md:235]: We are focusing on Linecard side for now, and sys log when the fabric capacity is not enough to sustain linerate for forwarding ASIC.  

We can extend some check in future for Fabric cards.

File: doc/voq/fabric.md [hunk 8]
```diff
@@ -169,7 +170,111 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+The design of fabric link monitor is intentionally scoped to use local component state such as information local to a linecard or information local to a supervisor. This design simplifies the need for inter-component communication.
+
+### 2.8.1 Monitor Fabric Link Status
+
+Unhealthy fabric links may lead to traffic drops. Fabric link monitoring is an important tool to minimize traffic loss. The fabric link monitor algorithm monitors fabric link status and isolates the link if one or more criteria are true. By isolating a fabric link, the link is still up in the physical layer, but is taken out of service and does not distribute traffic. This feature is needed on both fabric ASICs and forwarding ASICs.  
+
+#### 2.8.1.1 Fabric link monitoring criteria
+
+The fabric link monitoring algorithm checks two type of errors on a link: crc errors and uncorrectable errors. 
+
```
Comment (rlhui) [inline doc/voq/fabric.md:188]: Are these already defined in SAI port stats?  Can we list the exact sai_port_stat_t type being used for this?

File: doc/voq/fabric.md [hunk 9]
```diff
@@ -169,7 +170,111 @@ PORT            RxCells     TxCells      Crc       Fec  Corrected
...
+The criteria can be extended to include checking other errors later.
+
+#### 2.8.1.2 Monitoring algorithm
+
+Instead of reacting to the counter changes, Orchagent adds a new poller and periodically polls status of all fabric links. By default, the total number of received cells, cells with crc errors, cells with uncorrectable errors are fetched from all serdes links periodically and the error rates are calculated using these numbers. If any one of the error rates is above the threshold for a number of consecutive polls, the link is identified as an unhealthy link. Then the link is automatically isolated to not distribute traffic.
```
Comment (rlhui) [inline doc/voq/fabric.md:193]: In above "show fabric counters" sample output, it does not show "uncorrectable errors".  Can it be added?

Comment (jfeng-arista) [reply doc/voq/fabric.md:193]: This command output is not the new content that we added for monitoring algorithm. It is in the previous versions of document, the output might not up to date. 

Now I updated it with the output as of Aug.

File: doc/voq/fabric.md (hunk 8 above)
Comment (jfeng-arista) [reply doc/voq/fabric.md:188]: Yes. 
SAI_PORT_STAT_IF_IN_ERRORS is for crc erros 
SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES is for uncorrectable erros.

File: doc/voq/fabric.md (hunk 4 above)
Comment (arlakshm) [reply doc/voq/fabric.md]: @kenneth-arista , @jfeng-arista 

According to this document, currently if fabric links are bad the links are isolated automatically by SDK. Can you confirm this?

Comment (jfeng-arista) [reply doc/voq/fabric.md]: If the monitoring algorithm detected the link, and considered it as unhealthy, the algorithm calls sdk api to isolate the link, users do not need issue any CLI.

**Comment by kenneth-arista** [comment]
@rlhui and @arlakshm, this PR has already been reviewed by the community. If there are no further comments, can we merge this?

File: doc/voq/fabric.md [hunk 10]
```diff
@@ -155,21 +156,121 @@ When a forwarding ASIC is initialized, the fabric ports are initialized by defau
...
+```
+> config fabric port isolate [port_id]
+```
+
+```
+> config fabric port unisolate [port_id]
+```
+
+Besides the fabric link monitoring algorithm, the above two commands are added. The commands can be used to manually isolate and unisolate a fabric link ( i.e. take the link out of service and put the link back into service ). The two commands can help us debug on the system as well as force isolate a fabric link. 
+
```
Comment (rlhui) [inline doc/voq/fabric.md:223]: can this command be done on supervisor and line cards?  which SAI attribute will be used for this?

In this SAI PR https://github.com/opencomputeproject/SAI/pull/1764, it is applicable for fabric switch only.

However, should we also need to isolate a link from line card , can this SAI attribute be used then?

File: doc/voq/fabric.md [hunk 11]
```diff
@@ -155,21 +156,121 @@ When a forwarding ASIC is initialized, the fabric ports are initialized by defau
...
--------------------------------------------------------------------------
- 0           : 71660578         2         0         0         0
- 1           : 71659798         1         0         0       213
- 2           :        0         1         0         0       167
- 3           :        0         2         0         0       193
+#### 2.8.1.1 Cli commands
+
+Several commands will be added to set fabric link monitor config parameters.
+```
+> config fabric port monitor error threshold <#crcCells> <#rxCells>
```
Comment (rlhui) [inline doc/voq/fabric.md:195]: Please add Yang table schema for all new configurations proposed.

File: doc/voq/fabric.md (hunk 10 above)
Comment (jfeng-arista) [reply doc/voq/fabric.md:223]: The command can be used on both supervisor and Linecards for isolate a fabric port.  SAI_PORT_ATTR_FABRIC_ISOLATE will be used for this. 

https://github.com/opencomputeproject/SAI/pull/1764 is different and that is NOT for link level isolation.

Comment (rlhui) [reply doc/voq/fabric.md:223]: which SAI attribute is needed to do this fabric port isolation then? Thanks.

Comment (jfeng-arista) [reply doc/voq/fabric.md:223]: SAI_PORT_ATTR_FABRIC_ISOLATE

File: doc/voq/fabric.md (hunk 11 above)
Comment (jfeng-arista) [reply doc/voq/fabric.md:195]: yes, will add the Yang table for all the new configurations.

**Comment by kenneth-arista** [comment]
@rlhui can we get an approving review since all comments have been addressed?

**Review by arlakshm — APPROVED** [review]

**Comment by gechiang** [comment]
@jfeng-arista  There are many PRs to be picked up for this feature.  Since each PR approval run at is own timeline, can you share if there are any dependency that we need to be aware of?  Example of dependency concerns are something such as certain PR needs to go in first or else it may cause:
1. build time issue
2. run time dependency issue regardless of feature is being used

If there are dependency issue, please specify them clearly in this PR as well as in the corresponding PR for the dependency...
Thanks!

File: doc/voq/fabric.md [hunk 12]
```diff
@@ -155,35 +163,369 @@ When a forwarding ASIC is initialized, the fabric ports are initialized by defau
...
+
+#### 2.8.1.1 Fabric link monitoring criteria
+
+The fabric link monitoring algorithm checks two type of errors on a link: crc errors and uncorrectable errors. 
+
+The criteria can be extended to include checking other errors later.
+
+#### 2.8.1.2 Monitoring algorithm
+
+Instead of reacting to the counter changes, Orchagent adds a new poller and periodically polls status of all fabric links. By default, the total number of received cells, cells with crc errors, cells with uncorrectable errors are fetched from all serdes links periodically and the error rates are calculated using these numbers. If any one of the error rates is above the threshold for a number of consecutive polls, the link is identified as an unhealthy link. Then the link is automatically isolated to not distribute traffic.
```
Comment (rlhui) [inline doc/voq/fabric.md:219]: Please confirm (add in doc) whether the automatic isolation of unhealthy link is in 202205 release

Comment (jfeng-arista) [reply doc/voq/fabric.md:219]: Software based fabric link/capacity monitoring is not committed for 202205. The support will merge to master.

**Pull request merged**